rust/lance-index/src/scalar/zonemap.rs

```diff
@@ -0,0 +1,1629 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+//! Zone Map Index
+//!
+//! Zone maps are a columnar database technique for predicate pushdown and scan pruning.
+//! They break data into fixed-size chunks called "zones" and maintain summary statistics
+//! (min, max, null count) for each zone. This enables efficient filtering by eliminating
+//! zones that cannot contain matching values.
+//!
+//! Zone maps are "inexact" filters - they can definitively exclude zones but may include
+//! false positives that require rechecking.
+//!
+//!
+use super::btree::TrainingSource;
+use crate::scalar::SargableQuery;
+use crate::Any;
+use datafusion::functions_aggregate::min_max::{MaxAccumulator, MinAccumulator};
+use datafusion_expr::Accumulator;
+use futures::TryStreamExt;
+use lance_core::cache::LanceCache;
+use lance_core::ROW_ADDR;
+use std::sync::LazyLock;
+
+use arrow_array::{new_empty_array, ArrayRef, RecordBatch, UInt32Array, UInt64Array};
+use arrow_schema::{DataType, Field};
+use datafusion::execution::SendableRecordBatchStream;
+use datafusion_common::ScalarValue;
+use std::{collections::HashMap, sync::Arc};
+
+use super::{AnyQuery, IndexStore, MetricsCollector, ScalarIndex, SearchResult};
+use crate::scalar::FragReuseIndex;
+use crate::vector::VectorIndex;
+use crate::{Index, IndexType};
+use async_trait::async_trait;
+use deepsize::DeepSizeOf;
+use lance_core::Result;
+use lance_core::{utils::mask::RowIdTreeMap, Error};
+use roaring::RoaringBitmap;
+use snafu::location;
+const ZONEMAP_DEFAULT_SIZE: usize = 8192; // 1 zone every two batches
+
+const ZONEMAP_FILENAME: &str = "zonemap.lance";
+const ZONEMAP_SIZE_META_KEY: &str = "zonemap_size";
+
+/// Basic stats about zonemap index
+#[derive(Debug, PartialEq, Clone)]
+struct ZoneMapStatistics {
+    min: ScalarValue,
+    max: ScalarValue,
+    null_count: u32,
```

Comment:
I think we should also have a nan_count for floats

Reply:
added